pilot/pkg/networking/core/cluster.go

```diff
@@ -184,17 +191,30 @@ func (configgen *ConfigGeneratorImpl) deltaFromDestinationRules(updatedDr model.
 			return nil, nil
 		}
 		dr := prevCfg.Spec.(*networking.DestinationRule)
-		services = append(services, proxy.SidecarScope.ServicesForHostname(host.Name(dr.Host))...)
+		if features.FilterGatewayClusterConfig && proxy.Type == model.Router {
+			services = append(services, push.GatewayServicesForHostname(proxy, host.Name(dr.Host), envoyFilterPatches)...)
+		} else {
+			services = append(services, proxy.SidecarScope.ServicesForHostname(host.Name(dr.Host))...)
+		}
```

Comment:
maybe just extract a function and call it here and in the other two places below ?